src/main/java/com/pylon/spokestack/microsoft/BingSpeechClient.java

```diff
@@ -0,0 +1,375 @@
+        String header =
+            "Path:speech.config\r\n"
+            + String.format("X-RequestId:%s\r\n", this.requestId)
+            + String.format("X-Timestamp:%s\r\n", now())
+            + "Content-Type:application/json; charset=utf-8\r\n";
+        String body = "{}";
+        String message = String.format("%s\r\n%s", header, body);
+        this.socket.send(message);
+    }
+
+    private void sendHeader() throws Exception {
+        ByteBuffer message = ByteBuffer.allocateDirect(44)
+            .order(ByteOrder.LITTLE_ENDIAN)         // wave endian
+            .put("RIFF".getBytes("US-ASCII"))       // riff chunk id
+            .putInt(0)                              // file size
+            .put("WAVE".getBytes("US-ASCII"))       // riff format
+            .put("fmt ".getBytes("US-ASCII"))       // begin format chunk
+            .putInt(16)                             // format chunk size
+            .putShort((short) 1)                    // format (1=PCM)
+            .putShort((short) 1)                    // channel count
+            .putInt(this.sampleRate)                // sample rate
+            .putInt(this.sampleRate * 2)            // byte rate
+            .putShort((short) 2)                    // block alignment
+            .putShort((short) 16)                   // bits per sample
+            .put("data".getBytes("US-ASCII"))       // begin data chunk
+            .putInt(0);                             // data chunk size
+        sendAudio(message);
+        flush();
+    }
+
+    /**
+     * transmits an audio frame over the websocket.
+     * @param frame the audio frame buffer to send
+     * @throws Exception on error
+     */
+    public void sendAudio(ByteBuffer frame) throws Exception {
+        if (this.socket == null)
+            throw new IllegalStateException();
+
+        if (frame == null || this.buffer.remaining() < frame.capacity())
+            flush();
+
+        if (this.buffer.position() == 0) {
+            byte[] header = (
+                "Path:audio\r\n"
+                + String.format("X-RequestId:%s\r\n", this.requestId)
+                + String.format("X-Timestamp:%s\r\n", now())
+                + "Content-Type:audio/x-wav\r\n"
+            ).getBytes("US-ASCII");
+
+            this.buffer
+                .order(ByteOrder.BIG_ENDIAN)        // protocol endian
+                .putShort((short) header.length)    // message header length
+                .put(header);                       // message header
+        }
+
+        if (frame != null) {
+            frame.rewind();
+            this.buffer.put(frame);                 // message body
+        }
+    }
+
+    /**
+     * sends an empty audio frame indicating the end of the turn.
+     * @throws Exception on error
+     */
+    public void endAudio() throws Exception {
+        if (this.socket == null)
+            throw new IllegalStateException();
+
+        sendAudio(null);
+        flush();
```

Comment:
so this will flush twice for end of turn?

Reply:
Yes, and it has to. You have to first send the end of the previous message, and then you have to send an empty message to indicate the end of audio.